### ISSUE TITLE
Performance of CassandraTransactionalKeyValue benchmark is not good

### ISSUE THREAD

**Issue by GOGOYAO** [yugabyte/yugabyte-db#1910]
Jira Link: [DB-2502](https://yugabyte.atlassian.net/browse/DB-2502)
I use 3 nodes to build the cluster and I use the benchmark tool provided by https://docs.yugabyte.com/latest/benchmark/performance/ to run CassandraTransactionalKeyValue. 

Every node has a 32-core CPU and has a centos 7.6 OS installed. I use 1 nvme ssd disk in every nodes.  Every node is a brand-new machine.

But I get a bad result. Only in the first few seconds, I can get more than 10 thousand throughput and then the throughput fluctuates a lot. Of course, cpu utilization is 100% only in the first few seconds and then fluctuates a lot. In this test, no nodes is dead. I wonder whether this result is expected or some error happened. 

The result is as follows. 

```
 |  Write: 13551.36 ops/sec (9.12 ms/op), 67800 total ops  |  Uptime: 5006 ms | maxWrittenKey: 67771 | maxGeneratedKey: 67938 |
 |  Write: 13128.45 ops/sec (9.52 ms/op), 133463 total ops  |  Uptime: 10007 ms | maxWrittenKey: 133163 | maxGeneratedKey: 133589 |
 |  Write: 11800.74 ops/sec (10.59 ms/op), 192478 total ops  |  Uptime: 15008 ms | maxWrittenKey: 192375 | maxGeneratedKey: 192605 |
 |  Write: 11215.40 ops/sec (11.14 ms/op), 248564 total ops  |  Uptime: 20009 ms | maxWrittenKey: 248470 | maxGeneratedKey: 248691 |
 |  Write: 10130.39 ops/sec (12.32 ms/op), 299223 total ops  |  Uptime: 25009 ms | maxWrittenKey: 299000 | maxGeneratedKey: 299349 |
 |  Write: 9558.10 ops/sec (13.06 ms/op), 347020 total ops  |  Uptime: 30010 ms | maxWrittenKey: 346689 | maxGeneratedKey: 347145 |
 |  Write: 8984.08 ops/sec (13.91 ms/op), 391947 total ops  |  Uptime: 35011 ms | maxWrittenKey: 391519 | maxGeneratedKey: 392073 |
 |  Write: 8023.78 ops/sec (15.60 ms/op), 432072 total ops  |  Uptime: 40012 ms | maxWrittenKey: 431912 | maxGeneratedKey: 432197 |
 |  Write: 8046.05 ops/sec (15.54 ms/op), 472307 total ops  |  Uptime: 45012 ms | maxWrittenKey: 472146 | maxGeneratedKey: 472432 |
 |  Write: 7521.22 ops/sec (16.57 ms/op), 509918 total ops  |  Uptime: 50013 ms | maxWrittenKey: 509497 | maxGeneratedKey: 510045 |
 |  Write: 7275.19 ops/sec (17.20 ms/op), 546299 total ops  |  Uptime: 55014 ms | maxWrittenKey: 546081 | maxGeneratedKey: 546424 |
 |  Write: 7317.69 ops/sec (17.06 ms/op), 582892 total ops  |  Uptime: 60014 ms | maxWrittenKey: 582385 | maxGeneratedKey: 583017 |
 |  Write: 6695.95 ops/sec (18.69 ms/op), 616376 total ops  |  Uptime: 65015 ms | maxWrittenKey: 616136 | maxGeneratedKey: 616501 |
 |  Write: 6349.71 ops/sec (19.66 ms/op), 648129 total ops  |  Uptime: 70016 ms | maxWrittenKey: 647901 | maxGeneratedKey: 648255 |
 |  Write: 6185.03 ops/sec (20.22 ms/op), 679058 total ops  |  Uptime: 75016 ms | maxWrittenKey: 678640 | maxGeneratedKey: 679184 |
 |  Write: 5996.02 ops/sec (20.71 ms/op), 709042 total ops  |  Uptime: 80017 ms | maxWrittenKey: 708730 | maxGeneratedKey: 709167 |
 |  Write: 6076.60 ops/sec (20.63 ms/op), 739429 total ops  |  Uptime: 85018 ms | maxWrittenKey: 738978 | maxGeneratedKey: 739554 |
 |  Write: 5560.96 ops/sec (22.50 ms/op), 767238 total ops  |  Uptime: 90018 ms | maxWrittenKey: 766854 | maxGeneratedKey: 767363 |
 |  Write: 5468.23 ops/sec (22.80 ms/op), 794583 total ops  |  Uptime: 95019 ms | maxWrittenKey: 793931 | maxGeneratedKey: 794709 |
 |  Write: 5556.21 ops/sec (22.57 ms/op), 822368 total ops  |  Uptime: 100020 ms | maxWrittenKey: 822034 | maxGeneratedKey: 822492 |
 |  Write: 3632.32 ops/sec (34.40 ms/op), 840532 total ops  |  Uptime: 105020 ms | maxWrittenKey: 840175 | maxGeneratedKey: 840657 |
 |  Write: 5421.72 ops/sec (22.97 ms/op), 867644 total ops  |  Uptime: 110021 ms | maxWrittenKey: 867238 | maxGeneratedKey: 867769 |
 |  Write: 3399.56 ops/sec (36.80 ms/op), 884644 total ops  |  Uptime: 115022 ms | maxWrittenKey: 884074 | maxGeneratedKey: 884768 |
 |  Write: 4219.05 ops/sec (23.78 ms/op), 905742 total ops  |  Uptime: 120022 ms | maxWrittenKey: 904419 | maxGeneratedKey: 905866 |
 |  Write: 4196.23 ops/sec (35.67 ms/op), 926726 total ops  |  Uptime: 125023 ms | maxWrittenKey: 926193 | maxGeneratedKey: 926854 |
 |  Write: 4970.29 ops/sec (25.20 ms/op), 951581 total ops  |  Uptime: 130024 ms | maxWrittenKey: 951064 | maxGeneratedKey: 951706 |
 |  Write: 4935.76 ops/sec (25.31 ms/op), 976263 total ops  |  Uptime: 135024 ms | maxWrittenKey: 975962 | maxGeneratedKey: 976387 |
 |  Write: 4893.21 ops/sec (25.40 ms/op), 1000732 total ops  |  Uptime: 140025 ms | maxWrittenKey: 1000140 | maxGeneratedKey: 1000856 |
 |  Write: 4469.20 ops/sec (25.87 ms/op), 1023081 total ops  |  Uptime: 145026 ms | maxWrittenKey: 1021688 | maxGeneratedKey: 1023205 |
 |  Write: 3184.21 ops/sec (42.23 ms/op), 1039004 total ops  |  Uptime: 150026 ms | maxWrittenKey: 1038524 | maxGeneratedKey: 1039132 |
 |  Write: 4907.73 ops/sec (25.55 ms/op), 1063546 total ops  |  Uptime: 155027 ms | maxWrittenKey: 1063048 | maxGeneratedKey: 1063671 |
 |  Write: 3249.60 ops/sec (38.61 ms/op), 1079796 total ops  |  Uptime: 160028 ms | maxWrittenKey: 1079558 | maxGeneratedKey: 1079922 |
 |  Write: 4825.80 ops/sec (25.83 ms/op), 1103929 total ops  |  Uptime: 165028 ms | maxWrittenKey: 1103626 | maxGeneratedKey: 1104056 |
 |  Write: 3148.35 ops/sec (39.64 ms/op), 1119673 total ops  |  Uptime: 170029 ms | maxWrittenKey: 1119166 | maxGeneratedKey: 1119798 |
 |  Write: 4792.70 ops/sec (26.03 ms/op), 1143640 total ops  |  Uptime: 175030 ms | maxWrittenKey: 1143013 | maxGeneratedKey: 1143764 |
 |  Write: 3085.24 ops/sec (40.48 ms/op), 1159068 total ops  |  Uptime: 180030 ms | maxWrittenKey: 1158703 | maxGeneratedKey: 1159192 |
 |  Write: 4635.25 ops/sec (27.18 ms/op), 1182247 total ops  |  Uptime: 185031 ms | maxWrittenKey: 1182007 | maxGeneratedKey: 1182373 |
 |  Write: 4159.41 ops/sec (29.42 ms/op), 1203047 total ops  |  Uptime: 190032 ms | maxWrittenKey: 1202084 | maxGeneratedKey: 1203172 |
 |  Write: 2852.86 ops/sec (44.75 ms/op), 1217313 total ops  |  Uptime: 195032 ms | maxWrittenKey: 1217164 | maxGeneratedKey: 1217437 |
 |  Write: 4506.05 ops/sec (27.70 ms/op), 1239846 total ops  |  Uptime: 200033 ms | maxWrittenKey: 1239588 | maxGeneratedKey: 1239970 |
 |  Write: 2950.46 ops/sec (41.91 ms/op), 1254600 total ops  |  Uptime: 205034 ms | maxWrittenKey: 1254223 | maxGeneratedKey: 1254726 |
 |  Write: 4251.53 ops/sec (29.59 ms/op), 1275863 total ops  |  Uptime: 210035 ms | maxWrittenKey: 1275519 | maxGeneratedKey: 1275987 |
 |  Write: 4339.48 ops/sec (28.91 ms/op), 1297563 total ops  |  Uptime: 215035 ms | maxWrittenKey: 1297184 | maxGeneratedKey: 1297689 |
 |  Write: 4235.69 ops/sec (29.30 ms/op), 1318744 total ops  |  Uptime: 220036 ms | maxWrittenKey: 1318018 | maxGeneratedKey: 1318868 |
 |  Write: 691.72 ops/sec (166.67 ms/op), 1322203 total ops  |  Uptime: 225036 ms | maxWrittenKey: 1321227 | maxGeneratedKey: 1322327 |
 |  Write: 972.48 ops/sec (96.06 ms/op), 1327066 total ops  |  Uptime: 230037 ms | maxWrittenKey: 1325590 | maxGeneratedKey: 1327190 |
 |  Write: 810.90 ops/sec (202.90 ms/op), 1331121 total ops  |  Uptime: 235038 ms | maxWrittenKey: 1330492 | maxGeneratedKey: 1331245 |
 |  Write: 4235.51 ops/sec (30.16 ms/op), 1352301 total ops  |  Uptime: 240038 ms | maxWrittenKey: 1351991 | maxGeneratedKey: 1352425 |
 |  Write: 5874.22 ops/sec (21.35 ms/op), 1381676 total ops  |  Uptime: 245039 ms | maxWrittenKey: 1381516 | maxGeneratedKey: 1381800 |
 |  Write: 6288.03 ops/sec (19.86 ms/op), 1413120 total ops  |  Uptime: 250040 ms | maxWrittenKey: 1412960 | maxGeneratedKey: 1413244 |
 |  Write: 5879.51 ops/sec (21.18 ms/op), 1442521 total ops  |  Uptime: 255040 ms | maxWrittenKey: 1442267 | maxGeneratedKey: 1442645 |
 |  Write: 5550.98 ops/sec (22.51 ms/op), 1470279 total ops  |  Uptime: 260041 ms | maxWrittenKey: 1469787 | maxGeneratedKey: 1470403 |
 |  Write: 5688.89 ops/sec (22.04 ms/op), 1498727 total ops  |  Uptime: 265041 ms | maxWrittenKey: 1498520 | maxGeneratedKey: 1498854 |
 |  Write: 5628.71 ops/sec (22.17 ms/op), 1526874 total ops  |  Uptime: 270042 ms | maxWrittenKey: 1526628 | maxGeneratedKey: 1526999 |
 |  Write: 5238.17 ops/sec (23.95 ms/op), 1553068 total ops  |  Uptime: 275043 ms | maxWrittenKey: 1552940 | maxGeneratedKey: 1553194 |
 |  Write: 5196.59 ops/sec (23.93 ms/op), 1579054 total ops  |  Uptime: 280043 ms | maxWrittenKey: 1578622 | maxGeneratedKey: 1579181 |
 |  Write: 4791.05 ops/sec (26.14 ms/op), 1603012 total ops  |  Uptime: 285044 ms | maxWrittenKey: 1602669 | maxGeneratedKey: 1603137 |
 |  Write: 4816.00 ops/sec (25.95 ms/op), 1627095 total ops  |  Uptime: 290044 ms | maxWrittenKey: 1626762 | maxGeneratedKey: 1627220 |
 |  Write: 4578.05 ops/sec (27.17 ms/op), 1649988 total ops  |  Uptime: 295045 ms | maxWrittenKey: 1649527 | maxGeneratedKey: 1650112 |
 |  Write: 4831.83 ops/sec (25.84 ms/op), 1674150 total ops  |  Uptime: 300046 ms | maxWrittenKey: 1673729 | maxGeneratedKey: 1674275 |
 |  Write: 4297.71 ops/sec (29.11 ms/op), 1695641 total ops  |  Uptime: 305046 ms | maxWrittenKey: 1695342 | maxGeneratedKey: 1695766 |
 |  Write: 4288.37 ops/sec (29.11 ms/op), 1717085 total ops  |  Uptime: 310047 ms | maxWrittenKey: 1716637 | maxGeneratedKey: 1717210 |
 |  Write: 4328.82 ops/sec (28.99 ms/op), 1738732 total ops  |  Uptime: 315047 ms | maxWrittenKey: 1738298 | maxGeneratedKey: 1738857 |
 |  Write: 4478.47 ops/sec (27.86 ms/op), 1761128 total ops  |  Uptime: 320048 ms | maxWrittenKey: 1760818 | maxGeneratedKey: 1761252 |
 |  Write: 3999.70 ops/sec (31.03 ms/op), 1781130 total ops  |  Uptime: 325049 ms | maxWrittenKey: 1780561 | maxGeneratedKey: 1781254 |
 |  Write: 4049.52 ops/sec (31.19 ms/op), 1801380 total ops  |  Uptime: 330049 ms | maxWrittenKey: 1801101 | maxGeneratedKey: 1801504 |
 |  Write: 3580.04 ops/sec (34.49 ms/op), 1819282 total ops  |  Uptime: 335050 ms | maxWrittenKey: 1818940 | maxGeneratedKey: 1819406 |
 |  Write: 3643.02 ops/sec (34.26 ms/op), 1837499 total ops  |  Uptime: 340051 ms | maxWrittenKey: 1837091 | maxGeneratedKey: 1837625 |
 |  Write: 3928.00 ops/sec (32.08 ms/op), 1857141 total ops  |  Uptime: 345051 ms | maxWrittenKey: 1856700 | maxGeneratedKey: 1857265 |
 |  Write: 3699.60 ops/sec (33.41 ms/op), 1875641 total ops  |  Uptime: 350052 ms | maxWrittenKey: 1875236 | maxGeneratedKey: 1875765 |
 |  Write: 3423.05 ops/sec (36.83 ms/op), 1892758 total ops  |  Uptime: 355052 ms | maxWrittenKey: 1892241 | maxGeneratedKey: 1892883 |
 |  Write: 3675.90 ops/sec (34.00 ms/op), 1911140 total ops  |  Uptime: 360053 ms | maxWrittenKey: 1910812 | maxGeneratedKey: 1911264 |
 |  Write: 3520.86 ops/sec (35.60 ms/op), 1928746 total ops  |  Uptime: 365053 ms | maxWrittenKey: 1928354 | maxGeneratedKey: 1928870 |
 |  Write: 3051.48 ops/sec (41.03 ms/op), 1944005 total ops  |  Uptime: 370054 ms | maxWrittenKey: 1943679 | maxGeneratedKey: 1944129 |
 |  Write: 3274.07 ops/sec (38.15 ms/op), 1960377 total ops  |  Uptime: 375054 ms | maxWrittenKey: 1959902 | maxGeneratedKey: 1960503 |
 |  Write: 3283.05 ops/sec (38.01 ms/op), 1976794 total ops  |  Uptime: 380055 ms | maxWrittenKey: 1976587 | maxGeneratedKey: 1976918 |
 |  Write: 3387.43 ops/sec (36.55 ms/op), 1993733 total ops  |  Uptime: 385055 ms | maxWrittenKey: 1993178 | maxGeneratedKey: 1993859 |
 |  Write: 3028.04 ops/sec (41.39 ms/op), 2008875 total ops  |  Uptime: 390056 ms | maxWrittenKey: 2008566 | maxGeneratedKey: 2008999 |
 |  Write: 1979.99 ops/sec (62.81 ms/op), 2018776 total ops  |  Uptime: 395056 ms | maxWrittenKey: 2018114 | maxGeneratedKey: 2018900 |
 |  Write: 2978.10 ops/sec (42.58 ms/op), 2033668 total ops  |  Uptime: 400057 ms | maxWrittenKey: 2033262 | maxGeneratedKey: 2033792 |
 |  Write: 3051.08 ops/sec (40.82 ms/op), 2048925 total ops  |  Uptime: 405057 ms | maxWrittenKey: 2048513 | maxGeneratedKey: 2049050 |
 |  Write: 2832.30 ops/sec (44.28 ms/op), 2063088 total ops  |  Uptime: 410058 ms | maxWrittenKey: 2062857 | maxGeneratedKey: 2063212 |
 |  Write: 3379.03 ops/sec (37.13 ms/op), 2079985 total ops  |  Uptime: 415059 ms | maxWrittenKey: 2079725 | maxGeneratedKey: 2080109 |
 |  Write: 3069.69 ops/sec (40.36 ms/op), 2095335 total ops  |  Uptime: 420059 ms | maxWrittenKey: 2094782 | maxGeneratedKey: 2095460 |
 |  Write: 3121.06 ops/sec (40.01 ms/op), 2110942 total ops  |  Uptime: 425060 ms | maxWrittenKey: 2110631 | maxGeneratedKey: 2111066 |
 |  Write: 2246.15 ops/sec (55.85 ms/op), 2122174 total ops  |  Uptime: 430060 ms | maxWrittenKey: 2121546 | maxGeneratedKey: 2122298 |
 |  Write: 2887.47 ops/sec (43.32 ms/op), 2136613 total ops  |  Uptime: 435061 ms | maxWrittenKey: 2136223 | maxGeneratedKey: 2136737 |
 |  Write: 2940.50 ops/sec (42.11 ms/op), 2151317 total ops  |  Uptime: 440061 ms | maxWrittenKey: 2150847 | maxGeneratedKey: 2151441 |
 |  Write: 2834.30 ops/sec (44.80 ms/op), 2165490 total ops  |  Uptime: 445062 ms | maxWrittenKey: 2165279 | maxGeneratedKey: 2165616 |
 |  Write: 2871.30 ops/sec (42.85 ms/op), 2179848 total ops  |  Uptime: 450062 ms | maxWrittenKey: 2179390 | maxGeneratedKey: 2179972 |
 |  Write: 2987.27 ops/sec (42.08 ms/op), 2194786 total ops  |  Uptime: 455063 ms | maxWrittenKey: 2194464 | maxGeneratedKey: 2194910 |
 |  Write: 2852.50 ops/sec (43.95 ms/op), 2209050 total ops  |  Uptime: 460063 ms | maxWrittenKey: 2208785 | maxGeneratedKey: 2209174 |
 |  Write: 2173.58 ops/sec (57.43 ms/op), 2219919 total ops  |  Uptime: 465064 ms | maxWrittenKey: 2219614 | maxGeneratedKey: 2220043 |
```

**Comment by ttyusupov** [comment]
Hi @GoGoYao could you, please share exact command line you use to start the cluster and start the workload. Also, how much RAM does each node have and if there any other processes running on nodes except system ones?

**Comment by GoGoYao** [comment]
I follow steps in https://docs.yugabyte.com/latest/deploy/manual-deployment/ and I start the workload as follows.
```
java -jar ./yb-sample-apps.jar  \
      --workload CassandraTransactionalKeyValue   \
      --nodes $ENDPOINTS             \
      --nouuid                       \
      --value_size 102400               \
      --num_threads_read  200          \
      --num_threads_write  40       \
      --num_unique_keys 1000000000
```
No other processes was running and every node has 125G RAM.

**Comment by kmuthukk** [comment]
hi @GoGoYao 

1) Can you also share how much read ops/sec you are getting when running the above workload. In the snippet you posted, it only shows the write ops/sec.

2) If CPU is already maxed out, then beyond a certain limit, increasing the number of reader/writer threads can make latencies worse. Can you try with lower number of threads for read and write.

**Comment by GoGoYao** [comment]
@kmuthukk

**Comment by GoGoYao** [comment]
So, is there some official result about CassandraTransactionalKeyValue? In the homepage of yugabyteDB, I see the benchmark result if only on  CassandraKeyValue.

**Comment by kmuthukk** [comment]
This could be a possible regression due to issue described in #1951.  Fix for #1951 should be coming within a day.

@GoGoYao - which release of YB were you using? 

CC: @rao-vasireddy , @spolitov .

**Comment by bmatican** [comment]
Hey @GoGoYao, we're working on getting more of our internal performance benchmark results out.  For now, these are some of the numbers we have from a 3-node GCP `n1-standard-16`, with 2 local SSDs:

API | Workload | Read/Write threads | Throughput | Latency | Avg CPU on DB node
-- | -- | -- | -- | -- | --
YCQL | CassandraTransactionalKeyValue | 0/32 | 5.7k writes | 5.56ms writes | 59%
YCQL | CassandraTransactionalKeyValue | 32/0 | 32.5K | 0.98ms reads | 40%
YCQL | CassandraTransactionalKeyValue Writes | 12/4 | 9.7K reads / 1.4k writes | 1.24ms reads / 2.8ms writes | 36%
YCQL | CassandraTransactionalKeyValue Index | 4/12 | 2.6k reads / 3.3k writes | 1.5ms reads / 3.58ms writes | 51%

Something to consider for your workload:
- you have 32cpus, so probably can expect 2x these numbers
- you should consider lowering the total number of threads in the sample app (as you can see in our testing, already on 32 writers, we were hitting 60% cpu)
- are you running the sample app from one of the nodes? -- as you're basically maxing out CPU, this will further limit resources on that node.

cc @kmuthukk @ttyusupov
